### DISCUSSION TITLE
New layout idea: ThreeColumns

### DISCUSSION THREAD

**Discussion by mbenford** [qtile/qtile#2261] [Ideas]
I've created a very basic custom layout based on `Columns` that displays up to 3 columns and makes the middle one wider than the others (inspired on xmonad's [ThreeColumns layout](https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Layout-ThreeColumns.html)). It's very useful when you have an ultrawide display and you want to keep the window you are working on right in front of your eyes:
 
![qtile-three-colum-layout](https://user-images.githubusercontent.com/446867/109078594-60921080-76dc-11eb-9dd8-220eeaaba515.gif)

It's just a proof of concept and some things need to be sorted out before I can even think of sending in a PR, and I'd like to discuss them with anyone interested in such a layout.

I myself have the following questions so far:

- Should the main window be splittable/stackable, like a regular column? Or should it only hold one window?
- When there are more than 3 windows open and the main window is closed, what window should take its place? The newest? The oldest?
- Should the main column be ignored when swapping windows? I think there must be a way to promote a side window to the main column (like monadtall's `swap_main`), but I'm not sure about freely moving windows in and out of it.

Any input is appreciated. :)

**Comment by floppyman** [comment]
I like your idea, and it is exactly what I have been looking for, for my transition into Linux, I currently do something similar (as close as possible) on Windows with Microsoft PowerToys.

Currently me screen is separated like this:
* Center = 43% width (there about)
* Left and Right = split the remaining width 50/50
* Space between windows = 10px (so width is -40px of total)

This feels to me as a good split of the ultrawide (21:9) width (5120), as the center has enough space to function for most/all things (browsing, coding, etc), and is not so wide that you feel the need to turn your head or strain your eyes moving from side to side of the window.

And the Left and Right sides are not to small to be unusable for most things, like chat, browsing, music, reference material, etc.

For me I would like to see the following options/config settings:
* 3 columns are always active (meaning first window will open in center, and maybe the next is to the right, then left, so they will open clock-wise, on the sides)
* Defining the width of the center column or just a percentage definition for all 3 columns adding up to 100% (user calculated), maybe if the config option is an array then it could be: 1 number = center width, sides split rest, 3 numbers = 1 for each column
* Window spacing in pixels (with handling of not adding double spacing between columns or rows)
* Window border width in pixels
* Window border colors (focus, inactive) hex or rgb

As to your questions:
**Should the main window be splittable/stackable, like a regular column? Or should it only hold one window?**
I think it should be a single window in the center, like a Master/Stack layout, as it is the window you want to be in focus, the one your are working on, so all other windows should be pushed to the side.

**When there are more than 3 windows open and the main window is closed, what window should take its place? The newest? The oldest?**
Maybe just the last opened (newest), could possibly be a config option for newest (default) or oldest.

**Should the main column be ignored when swapping windows? I think there must be a way to promote a side window to the main column (like monadtall's swap_main), but I'm not sure about freely moving windows in and out of it.**
I agree with doing a promotion option for moving a window to center column, and just allowing to move/swap windows around on the sides.

**Reply by mbenford** [reply]
Thanks for weighing in! I agree with your points. I'm not sure they all could be done by simply extending the existing Columns layout, as I did initially, and perhaps creating a new one from scratch might be needed.

I'll send a PR later with my current implementation as a starting point for further discussion.

**Comment by m-col** [comment]
Perhaps a way to set the ratios of column widths in the existing 
Columns layout would achieve this?

**Reply by floppyman** [reply]
True, it would solve some of it.

But could it also be possible to define a column as the Main/Master column, so that it will become a ThreeColumnMasterStack type of layout.

Maybe that is beyond the scope of the existing layout, as it will try to solve too many things.

**Reply by mbenford** [reply]
That is precisely what I did in my PoC:

```python
class ThreeColumns(layout.Columns):
    defaults = [
        ("main_size", 50, "")
    ]
    def __init__(self, **config):
        layout.Columns.__init__(self, **config)
        self.add_defaults(ThreeColumns.defaults)
        self.num_columns = 3

    def add(self, client):
        super().add(client)

        if len(self.columns) == 3:
            main_width = 300 * self.main_size / 100
            other_width = (300 - main_width) / 2
            self.columns[0].width = other_width
            self.columns[1].width = main_width
            self.columns[2].width = other_width
            self.group.layout_all()

    def remove(self, client):
        super().remove(client)

        if 0 < len(self.columns) < 3:
            for col in self.columns:
                col.width = 100
            self.group.layout_all()
```

But I'm not sure if it's possible to accomplish other things this way, like moving the oldest window to the main column when its current window is closed. I wonder if extracting some of the `Columns` code into a base class that could be used by both layouts would make things easier?

**Comment by hamensman** [comment]
I would certainly like having the center column holding a master window - more akin to 'CenteredMaster' layout of XMonad.

So, if you spawn one window, it takes up the whole screen (as with MonadTall and MonadWide). A second window, the first (master) window is on the left and the second (slave) widow on the right (again, consistent with MonadTall) A third window, that's when you would see that first window come in the center, as the third would be on the left. A fourth, not sure how XMonad's CenteredMaster behaves, but I would guess that would be under the third (on the left side). A fifth, under the second (right side). A sixth, maybe under the fourth (left side)? A seventh, under the firth (right side)?

However CenteredMaster on XMonad behaves, go for that.

**Reply by mbenford** [reply]
[CenteredMaster](https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Layout-CenteredMaster.html) isn't actually a layout. It modifies another layout and centers the main window on top of the everything else. The XMonad layout that does exactly what you described is [ThreeColumns](https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Layout-ThreeColumns.html), which is the one I plan to port to qtile.

I think you meant dwm's [CenteredMaster patch](https://dwm.suckless.org/patches/centeredmaster), which seems to behave like XMonad's ThreeColMid layout.

**Reply by hamensman** [reply]
I think it would also be cool if your layout can allow more windows in the (centre) master stack, and also dictate their orientation (whether the master stack**s** are tiled vertically or horizontally). And having keybinds to change the number of master stacks and their orientation on the fly.

**Comment by mbenford** [comment]
I submitted a PR with a functional port of XMonad's ThreeColumns layout: #2606.
